test/jdk/java/util/IdentityHashMap/Basic.java

```diff
@@ -38,33 +43,84 @@
  * @run testng Basic
  */
 
-// TODO need this? @modules java.base/java.util:open
-// TODO add tests using null
-// TODO copy constructor instead of k3 and v3 (key and value not in map)?
+// NOTE: avoid using TestNG's assertEquals/assertNotEquals directly on two Map instances,
+// as its logic for testing collections equality is suspect. Use checkContents() to assert
+// that a map's entrySet contains exactly the expected mappings.
+
+// TODO remove(k, v)
+// TODO replace(k, v1, v2)
+// TODO add tests using null keys and values
+// TODO view collections iterators
+// TODO Map.Entry::setValue
 
 public class Basic {
+    /*
+     * Helpers
+     */
+
     record Box(int i) {
         Box(Box other) {
             this(other.i());
         }
     }
 
-    Predicate<Box> isIdenticalBox(Box b1) {
-        return b2 -> b1 == b2;
+    // Checks that a collection contains exactly the given elements and no others, using the
+    // provided predicate for equivalence. Checking is performed both using contains() on the
+    // collection and by simple array searching. The latter is O(N^2) so is suitable only for
+    // small arrays. No two of the given elements can be equivalent according to the predicate.
+
+    @SafeVarargs
+    private <E> void checkContents(Collection<E> c, BiPredicate<E,E> p, E... given) {
+        @SuppressWarnings("unchecked")
+        E[] contents = (E[]) c.toArray();
```

Comment:
Maybe worthy to note that entry set toArray may return entries different from what the iterator returns, and some predicates may thus fail.

Reply:
Yeah I could add a note to caution against checking identity of `Entry` instances. That doesn't occur anywhere in these tests, does it?